Bug fixes and improvements

english5040 opened the pull request:
A lot of code changes here, so this probably needs some testing

english5040 committed:
Pop Star: fixes

`v.debuff == false` may give an incorrect result if `v.debuff` is nil.

Use poll_enhancement. This is good style, but
also avoids a bug with randomly generating a special enhancement that
another mod declared with `in_pool() == false`. Most "pools" of items
in Balatro shouldn't be directly selected from.

english5040 committed:
Use SMODS probability funcs

These are required now for mod compatibility:
probability-modifying effects and triggering on probabilities
in other mods use these functions

english5040 committed:
Fix remaining calls to pseudorandom()

The way pseudorandom() was used is incorrect,
the right way is to directly call psuedorandom() once.
Using `math.random()` is wrong because pseudorandom() internally
takes care of seeding logic. `math.random()` ignores this system,
so it'll actually ignore a Balatro run's starting seed.

english5040 committed:
Fix possible nil

english5040 committed:
Frequent Flyer doesn't trigger on buying itself

english5040 committed:
Better and shorter Handheld is_suit hook

Maintains the exact same if-then logic, but
won't clobber other mods' hooks (by reusing the
original function)

also: comment on description of Handheld: it says
cards are Wild but they're not actually treated
as Wild for e.g. Extra Credit's Werewolf

english5040 committed:
Replace comparisons for Talisman mod compatibility

Talisman is a mod that uncaps numbers past 1e308, by replacing them
with tables (big numbers) that have arithmetic operations overloaded.

Unfortunately Lua still can't compare bignums with normal numbers so
comparisons with game values need both sides wrapped with `to_big()`
to convert. (If Talisman isn't loaded, `to_big()` is a no-op)

english5040 committed:
Fix Dunce: if there is no joker to right, sets next_joker to nil

(next_joker should just always be set)

english5040 committed:
Make as many global variables local as possible

Effigy needed to save a value in the Joker itself.
I've shown the usual method for doing this.

Some jokers required global variables and their code
could not be easily changed. Added comments to those

Using global variables in a Joker is almost always wrong
and will lead to these bugs:
- Multiples of that Joker clobber each other's variables
- State persists after starting a new run, and wrongly does not persist after loading a saved game
- Will conflict with other mods

english5040 committed:
Add comment on Frequent Flyer

english5040 commented:
Still working on reviewing the code

BrandonE commented:
Please provide an explanation for the purpose of each change. Several of these look like stylistic changes and not bug fixes. I would like to understand which of these changes impacts the gameplay and how. I also don't know anything about the Talisman mod and would like to get some context there.

english5040 commented:
fair, I shouldn't assume any background knowledge and I can certainly change the commit messages to explain in detail. sorry I am working on other stuff in the meantime

english5040 committed:
Fix typos in localization

english5040 committed:
Trash localization/default.lua

localization/en-us.lua is already used as a
fallback by SMODS. Only one of the two is needed.

english5040 committed:
Comment on Effigy's description

Effigy's description doesn't accurately describe it.
Effigy also has a bug: it will have strange timing with
context.before/context.after effects like Burnt Joker:
depends on if Effigy was placed before or after that Joker.

english5040 committed:
Fix Pyromaniac

The global variable was wrong, and also burn_hand was not reset, so Pyromaniac destroyed all hands played

english5040 committed:
Enhancement-related jokers should be gated

These should only show up if you have at least one of that enhancement

english5040 committed:
Scaling Jokers can't be perishable.

This is true in the base game.

(Unstoppable Force is ok, it's not technically the same joker)

english5040 committed:
Intuition only shows up if you have a Steel or Gold Card
